src/functional/merge.js

```diff
@@ -68,11 +72,23 @@ export function mergeWithSources(collection, sources, merger) {
 
 function deepMergerWith(merger) {
   function deepMerger(oldValue, newValue, key) {
-    return isDataStructure(oldValue) && isDataStructure(newValue)
+    return isDataStructure(oldValue) &&
+      isDataStructure(newValue) &&
+      areMergeable(oldValue, newValue)
       ? mergeWithSources(oldValue, [newValue], deepMerger)
       : merger
       ? merger(oldValue, newValue, key)
       : newValue;
   }
   return deepMerger;
 }
+
+function areMergeable(oldDataStructure, newDataStructure) {
+  const oldSeq = Seq(oldDataStructure);
+  const newSeq = Seq(newDataStructure);
+  return (
+    isIndexed(oldSeq) === isIndexed(newSeq) &&
+    isKeyed(oldSeq) === isKeyed(newSeq) &&
+    isSet(oldSeq) === isSet(newSeq)
```

Comment:
You may need some test coverage to cover this, but I believe `isSet` only checks if something is actually `Set` or `OrderedSet` and doesn't include `SetSeq`. We don't have a great first class function for this, something is set-like if it is both not indexed and not keyed

Reply:
I wondered about that, I should've looked into that more, makes sense.